### PR TITLE
Fix baremetal implementation and reorganize things

### PR DESCRIPTION
**Pull request by linkmonitor**
This PR has two high-level goals:

1. Fix issues in the Baremetal implementation identified by Fumiaki Kagaya.
2. Reorganize how examples are written and add unit tests.

All tests and example applications are built simultaneously and the sources live under `extra/`.

### CODE REVIEW

File: eventrouter/internal/eventrouter_baremetal.c [hunk 1]
```diff
@@ -57,7 +58,8 @@ void ErInit(const ErOptions_t *a_options)
 
     for (size_t idx = 0; idx < task->m_num_modules; ++idx)
     {
-        ErModule_t *module   = task->m_modules[idx];
+        ErModule_t *module = task->m_modules[idx];
```
Comment (linkmonitor) [inline eventrouter/internal/eventrouter_baremetal.c:61]: `clang-format` change.

File: eventrouter/internal/eventrouter_baremetal.c [hunk 2]
```diff
@@ -207,14 +207,17 @@ void ErUnsubscribe(ErModule_t *a_module, ErEventType_t a_event_type)
...
 {
-    /// Keep events which may not have been delivered during the previous loop
-    /// at the head of the "deliver now" list and add events which were
-    /// scheduled for delivery during the previous loop.
-    ErListAppend(&s_context.m_events.m_deliver_now,
-                 s_context.m_events.m_deliver_next.m_next);
-    /// Clear the "deliver next" list so it can be filled during this loop and
-    /// delivered during the next loop.
-    s_context.m_events.m_deliver_next.m_next = NULL;
+    if (s_context.m_events.m_deliver_next.m_next != NULL)
```
Comment (linkmonitor) [inline eventrouter/internal/eventrouter_baremetal.c:210]: A silly mistake; the program crashed if there was a loop where no events were generated.

File: eventrouter/internal/eventrouter_baremetal.c [hunk 3]
```diff
@@ -163,19 +165,17 @@ void ErReturnToSender(ErEvent_t *a_event)
...
     a_event->m_reference_count--;
 
-    if (a_event->m_reference_count > 1)
+    if (a_event->m_reference_count > 0)
     {
         // Do nothing. Some modules have KEPT the event and must explicitly
         // call `ErReturnToSender()` before we can return it.
     }
-    else if (a_event->m_reference_count == 1)
+    else if (a_event->m_reference_count == 0)
```
Comment (linkmonitor) [inline eventrouter/internal/eventrouter_baremetal.c:173]: An embarrassing error: it does not appear that events were returned to the sender correctly before this.

**Review by fumiaki-kagaya-yohana — APPROVED** [review]
Looks great! Thank you!!

**Pull request merged**